### ISSUE TITLE
[YW] [K8S] Edit Universe failed with RPC retry

### ISSUE THREAD

**Issue by ryli17** [yugabyte/yugabyte-db#3873]
Yugaware version: 2.1.0.0-b18
Yugabyte version: 2.1.0.0-b18

When we edit a 3-node universe from 3-region to 1-region, task failed with the following error:

```
Failed to execute task java.util.concurrent.FutureTaskffb332a71, hit error java.lang.RuntimeException:
org.yb.client.NonRecoverableException: Too many attempts: YRpc(method=6etMasterClusterConfig, service=yb.master.MasterService,
tablet=null, attempt=12, maxAttempts=100, DeadlineTracker(timeout=120000, elapsed=117523), Deferred@2024698768-(state=PENDING,
result=null, callback=(continuation of Deferredp1757445300 after retry RPC after error@949039450) -> (continuation of
Deferred@1061145696 after retry RPC after erronP872811236) -> (continuation of Deferred01598616544 after retry RPC after
errorip1860306014) -> (continuation of Deferred@796887342 after retry RPC after errorg24236047) (continuation of Deferred069435379
 after retry RPC after error,K307358499) -> (continuation of OeferredE603931721 after retry RPC after error11659337992) (continuation
 of Deferred0168999352 after retry RPC after errong832940419) -> (continuation of Deferred@939935305 after retry RPC after
errorP1632662396) -> (continuation of DeferredK212779693 after retry RPC after error@553992524) -> (continuation of Deferredki7511113813
 after retry RPC after error,K324106147) -› (continuation of Deferredek1246003255 after retry RPC after errorg992967921), errback=
 (continuation of Deferredg1757445300 after retry RPC after error@949039450) -> (continuation of Deferred@l061145696 after retry RPC
after error.g.72811236) -> (continuation of DeferredK598616&44 after retry RPC after errol1.1860306014) -> (continuation of
Deferred 796687342 after retry RPC after error@724236047) -> (continuation of Deferred@869435379 after retry RPC after errorK307358499)
-> (continuation of Deferre01503931721 after retry RPC after errorK659337992) -> (continuation of Deferredp1168999352 after retry RPC
after error-0532940419) -> (continuation of Deferred039935305 after retry RPC after enrol-D1632662396) -> (continuation of
 Deferred@1212779693 after retry RPC after errorip653992524) -> (continuation of Deferred0751111380 after retry RPC after
errorrg324106147) -> (continuation of DeferredK246003255 after retry RPC after error@902967921))).
```

Aha! Link: https://yugabyte-test.aha.io/features/PLATFORM-622

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-issue-3873-analysis.md`):

# Analysis: yugabyte/yugabyte-db#3873 — [YW] [K8S] Edit Universe failed with RPC retry

Validated 2026-08-30 against local checkout 46a9f7adfd (master) and tag v2.1.0
(the reported 2.1.0.0-b18 era). Sibling reports (same reporter, same incident,
2020-03-06): #3870 (retry accumulates pods), #3874 (WaitForDataMove failure +
data loss on retry). No Jira cross-link on the issue (Aha! PLATFORM-622 only).

## Verdict

NEEDS-INFO — the failure *mechanism* is fully characterized from 2.1.0.0-b18
source (a 120 s no-leader retry loop inside YBA's `UpdatePlacementInfo` subtask
immediately after the K8s master-quorum move), but the report contains zero
diagnostics (no YBA `application.log`, no yb-master logs, no task detail JSON,
OCR-mangled trace) and the reporter's environment (Yugaware 2.1.0.0-b18,
helm 2/Tiller-era Kubernetes) is ~6.5 years EOL, so the specific event that
left the quorum leaderless cannot be pinned to one of the identified candidate
defects. Every structural gap that enabled this failure class was subsequently
fixed in YBA (lineage below); the report is not actionable on current master.

## Root cause

**Failure mechanism (proven from v2.1.0 source):**

The reported signature — `NonRecoverableException: Too many attempts:
YRpc(method=GetMasterClusterConfig, service=yb.master.MasterService,
tablet=null, attempt=12, maxAttempts=100, DeadlineTracker(timeout=120000,
elapsed=117523), ... (continuation of Deferred… after retry RPC after error…)`
— is uniquely produced by this chain (all v2.1.0 citations):

- The only production caller of `getMasterClusterConfig()` in Yugaware 2.1.0 is
  the `UpdatePlacementInfo` subtask:
  `managed/src/main/java/com/yugabyte/yw/commissioner/tasks/subtasks/UpdatePlacementInfo.java:86-89`
  → `java/yb-client/src/main/java/org/yb/client/AbstractModifyMasterClusterConfig.java:34`,
  wrapped in the `RuntimeException` seen in the trace
  (`UpdatePlacementInfo.java:92`). In `EditKubernetesUniverse.run()` it is
  scheduled at
  `managed/src/main/java/com/yugabyte/yw/commissioner/tasks/EditKubernetesUniverse.java:146`,
  immediately after the master quorum move
  (`createMoveMasterTasks`, `EditKubernetesUniverse.java:199-215`: all
  ChangeConfig ADDs, then all REMOVEs, then one point-in-time
  `WaitForMasterLeader`) and the tserver scale-up.
- The subtask builds a **fresh** `YBClient` bootstrapped from
  `universe.getMasterAddresses()` (`UpdatePlacementInfo.java:79-84`) with
  `defaultAdminOperationTimeoutMs(120000)`
  (`managed/src/main/java/com/yugabyte/yw/common/services/LocalYBClientService.java:43-47`)
  — the trace's `timeout=120000`.
- For any master RPC, the 2.1.0 Java client fans `GetMasterRegistration` out to
  its fixed bootstrap list and succeeds only if some address answers
  role==LEADER
  (`java/yb-client/src/main/java/org/yb/client/GetMasterRegistrationReceived.java:181`).
  Follower answers are discarded (`:201-202`); an all-non-leader round raises
  `NoLeaderMasterFoundException` (`:144-146`), which the retry errback
  (`java/yb-client/src/main/java/org/yb/client/AsyncYBClient.java:1152-1159`,
  `toString()` = "retry RPC after error", `:1168` — the OCR-mangled
  continuation chain in the trace) converts into a delayed retry with linear
  backoff (`SLEEP_TIME=500`, `AsyncYBClient.java:157`; `sleepTime =
  attemptCount * SLEEP_TIME + rand(50)`, `:1288`) until the deadline fires
  `tooManyAttemptsOrTimeout` → "Too many attempts:" (`:1366-1372`).
- Arithmetic check: 12 attempts ⇒ ≈33 s cumulative backoff; the remaining
  ≈84 s of `elapsed=117523` is per-round fan-out latency waiting on
  unresponsive/wedged endpoints — consistent with a bootstrap list still
  containing the two just-removed regional masters.

Therefore the incident state is unambiguous: **for the full 120 s window, no
address in YBA's recorded master list answered "I am LEADER"** — the actual
master Raft quorum was leaderless, or the actual leader was absent from YBA's
list. (The signature also *excludes* two neighbouring modes: total DNS
resolution failure of all addresses would have produced the non-retriable
"Couldn't find a valid master" error, `GetMasterRegistrationReceived.java:147-151`,
and a reachable-but-bogus leader would have let the RPC succeed — cf. sibling
#3874's data loss.)

**Enabling defects in the 2.1.0 edit flow (why this state was reachable):**

- No verification that the quorum move completed/converged before the next
  master RPC: `ChangeMasterConfig.run()` checks only the immediate RPC response
  (`managed/.../subtasks/ChangeMasterConfig.java:94-118`); `WaitForMasterLeader`
  is a single point-in-time poll (`subtasks/WaitForMasterLeader.java:55-70`).
- YBA's master address list is *inferred from pod names*, wholesale-replacing
  `nodeDetailsSet` on every POD_INFO
  (`managed/.../subtasks/KubernetesCommandExecutor.java:300-344`; any pod whose
  hostname contains "master" gets `isMaster=true`, `:319-323`), so the two
  Raft-removed masters stay in `universe.getMasterAddresses()`
  (`managed/.../models/Universe.java:500-506,551-578`) until their helm
  releases are deleted much later (`EditKubernetesUniverse.java:169-173`), and
  the list is never reconciled against the actual Raft config.
- `UpdatePlacementInfo` had no retries (TODO at `UpdatePlacementInfo.java:98`),
  and the task was not re-runnable — retrying the whole edit produced the
  sibling failures (#3870 pod accumulation, #3874 data loss).

**Missing evidence (what NEEDS-INFO means here):** discriminating *which*
event de-leadered the quorum (a swallowed/partial ChangeConfig failure; new
masters added but wedged/unreachable when the removals landed; a stale
YBA-recorded master set) requires artifacts never attached to the issue and
unobtainable today: (1) yb-master INFO/WARNING logs from all five master pods
covering the edit window, (2) Yugaware `application.log` with the failed task's
subtask trail (which subtask emitted the trace, and the ChangeConfig/
WaitForMasterLeader outcomes preceding it), (3) the universe placement JSON
before/after the edit (regions/AZs, nodePrefix, per-AZ helm releases). Without
these, per-defect attribution is unprovable; with them it would be mechanical.

**Master still affected?** The failure *class* is closed on master by a chain
of later hardening (verified in the local checkout at 46a9f7adfd):
post-ChangeConfig quorum verification via `listMasterRaftPeers`
(`managed/.../subtasks/ChangeMasterConfig.java:218`; PLAT-9607 `23dc5fcbfa`,
PLAT-10162 `546a4872a8`, PLAT-15227 `db08e305aa`), unswallowed change-config
exceptions (PLAT-7236 `e8e0552e66`), YBA-vs-reality assignment verification
(`managed/.../subtasks/CheckClusterConsistency.java:173`; PLAT-10402
`b5d7386745`), retries for the cluster-config write (current
`AbstractModifyMasterClusterConfig` retries CONFIG_VERSION_MISMATCH; PLAT-7269
`ea5bebf502`), cluster-membership safety on pod delete (PLAT-11233
`17f7a183ec`), re-runnable K8s tasks (PLAT-13388 `55b167128e`), and an
explicit, guarded K8s full-move flow with rollback
(`EditKubernetesUniverse.java:383-529` on master; PLAT-18885+ `ba48df78b5`,
2026 rollback series PLAT-21483/21484/21485/21486 + `ConfirmEditRollbackMembership.java:74`).
The *client* behaviour (leader-required lookup, retry to deadline) persists by
design (`GetMasterRegistrationReceived.java:189` on master) — it is correct
client behaviour against a leaderless quorum; the orchestration that pointed it
at one mid-edit is what was fixed.

## Trigger matrix

Source-derived predictions for the 2.1.0 client facing `GetMasterClusterConfig`;
rows 1-2 executed on a current image (mechanism demo — see Reproduction), rows
3-4 are discriminated by the reported trace itself. A faithful end-to-end
replay of the 2020 incident is not executable (see Reproduction).

| # | condition (client bootstrap list vs quorum) | predicted outcome (v2.1.0 source) | observed |
|---|---|---|---|
| 1 | List contains the actual leader, healthy quorum (control) | Immediate success, config returned (`GetMasterRegistrationReceived.java:181-196`) | Executed: same probe returns cluster-config JSON in 0.16 s (`/tmp/claude/yb-issue-3873-control-demo.log`) |
| 2 | No listed master answers LEADER (leaderless quorum / leader outside list) | `NoLeaderMasterFoundException` every round → delayed retries → `Too many attempts` at deadline (`AsyncYBClient.java:1152-1159,1366-1372`) | Executed: probe retries for its full 15 s deadline, fails "Could not locate the leader master"; master log shows continuous `no leader found: GetLeaderMasterRpc(...)` (`/tmp/claude/yb-issue-3873-noleader-demo.log`) — matches the reported 120 s/attempt=12 shape |
| 3 | All listed addresses unresolvable | Non-retriable `NonRecoverableException "Couldn't find a valid master"` immediately (`GetMasterRegistrationReceived.java:123-151`) | Not executed; the reported trace (12 retries over 117.5 s) proves this was NOT the incident mode |
| 4 | A listed master claims LEADER of a wrong/bogus quorum | RPC succeeds against wrong leader; no timeout (failure surfaces later) | Not executed; consistent with sibling #3874 (data-move failure, data loss), not with #3873's timeout |

## Reproduction

**Faithful end-to-end reproduction does not apply.** The incident is a
race-dependent, one-shot orchestration failure inside Yugaware 2.1.0.0-b18
driving a helm 2/Tiller-era multi-region Kubernetes universe through a
3-region → 1-region edit (reporter's own retry produced *different* failures,
#3870/#3874). Standing up YW 2.1.0.0-b18 + Tiller + a 3-region K8s provider is
impractical and non-deterministic by nature; the protocol's
historical/current-status path was used instead. Historical trigger as
reported: create a 3-node RF3 K8s universe across 3 regions on YW/YB
2.1.0.0-b18, then Edit Universe to 1 region; the task fails at the
`UpdatePlacementInfo` step with the `GetMasterClusterConfig` retry-to-deadline
trace when the just-moved master quorum has no reachable leader in YBA's
recorded master list.

**Executed mechanism demo** (both halves of the failure state; podman,
aarch64, image `docker.io/yugabytedb/yugabyte:2026.1.1.1-b2-aarch64`):

```sh
# A) leaderless quorum: 1 of 3 configured masters running => can never win election
podman run -d --name yb3873-noleader docker.io/yugabytedb/yugabyte:2026.1.1.1-b2-aarch64 \
  bash -c "mkdir -p /tmp/m && /home/yugabyte/bin/yb-master --fs_data_dirs=/tmp/m \
    --master_addresses=127.0.0.1:7100,127.0.0.2:7100,127.0.0.3:7100 \
    --rpc_bind_addresses=127.0.0.1:7100 --webserver_interface=127.0.0.1 \
    --webserver_port=7000 --logtostderr"
sleep 8
# cluster-config fetch (yb-admin get_universe_config = GetMasterClusterConfig content)
time podman exec yb3873-noleader /home/yugabyte/bin/yb-admin \
  -master_addresses 127.0.0.1:7100 -timeout_ms 15000 get_universe_config
#  -> retries for the FULL 15 s deadline, then:
#     "Timed out ... Unable to establish connection to leader master"
#     ": Could not locate the leader master: Call timed out"
#  master's own log: "no leader found: GetLeaderMasterRpc(addrs: [...], num_attempts: 52)"

# B) control: single-master quorum => leader exists in the list
podman run -d --name yb3873-leader docker.io/yugabytedb/yugabyte:2026.1.1.1-b2-aarch64 \
  bash -c "mkdir -p /tmp/m && /home/yugabyte/bin/yb-master --fs_data_dirs=/tmp/m \
    --master_addresses=127.0.0.1:7100 --rpc_bind_addresses=127.0.0.1:7100 \
    --webserver_interface=127.0.0.1 --webserver_port=7000 --logtostderr"
sleep 8
time podman exec yb3873-leader /home/yugabyte/bin/yb-admin \
  -master_addresses 127.0.0.1:7100 -timeout_ms 15000 get_universe_config
#  -> 0.16 s: {"version":0,"clusterUuid":"...","universeUuid":"..."}

podman rm -f yb3873-noleader yb3873-leader
```

Failure signature to grep for (demo): `Could not locate the leader master` /
`no leader found: GetLeaderMasterRpc`. Reported Java-client equivalent:
`Too many attempts: YRpc(method=GetMasterClusterConfig` with
`DeadlineTracker(timeout=120000`. Logs preserved at
`/tmp/claude/yb-issue-3873-noleader-demo.log` and
`/tmp/claude/yb-issue-3873-control-demo.log`.

## Versions tested

- **v2.1.0 tag (2.1.0.0-b18 era) — source analysis**: full defect-path walk of
  `EditKubernetesUniverse`, `UpdatePlacementInfo`, `ChangeMasterConfig`,
  `WaitForMasterLeader`, `KubernetesCommandExecutor`, `Universe`,
  `LocalYBClientService`, yb-client (`AsyncYBClient`,
  `GetMasterRegistrationReceived`, `YBClient`,
  `AbstractModifyMasterClusterConfig`); helm chart `yugabyte/charts` branch 2.1
  (`stable/yugabyte/templates/service.yaml:273-277` — multi-AZ masters get
  `--master_addresses={{ .Values.masterAddresses }}`, so YBA's shell-mode `""`
  boots new masters with an empty master list, and STS partition shields
  running pods).
- **2.1.8_FINAL**: `EditKubernetesUniverse.java` unchanged from v2.1.0 (git
  diff empty) — all of 2.1.x affected.
- **master @ 46a9f7adfd (2026-08)**: failure-enabling gaps closed (citations in
  Root cause); leader-required client lookup persists by design.
- **docker.io/yugabytedb/yugabyte:2026.1.1.1-b2-aarch64** — executed
  leaderless/control mechanism demo (matrix rows 1-2).
- 2.1.0.0-b18 binaries were not executed (no aarch64 images exist for 2.1;
  end-to-end YW-on-K8s replay impractical — see Reproduction).

## Proposed fix direction

No code change proposed. The enabling defects are already fixed on master
(PLAT-7236, PLAT-7269, PLAT-9607, PLAT-10162, PLAT-10402, PLAT-11233,
PLAT-13388, PLAT-15227, PLAT-18885+, PLAT-21483..21486 — commit SHAs in Root
cause). Recommend closing #3873 as addressed-by-lineage (together with siblings
#3870/#3874, which are the retry-side symptoms of the same 2.1.0 edit-flow
non-idempotency), or, if maintainers want reporter confirmation first, asking
for: (1) yb-master logs from all master pods for the edit window, (2) Yugaware
application.log with the failed task's subtask trail, (3) the placement JSON
before/after the edit — none of which are likely recoverable in 2026.

## Test strategy

Covered by existing YBA unit tests for the modern flow:
`managed/src/test/java/com/yugabyte/yw/commissioner/tasks/EditKubernetesUniverseTest.java`
(subtask-sequence assertions incl. master-move + placement update) and
`.../subtasks/CheckClusterConsistencyTest.java` /
`EditUniverseTest`-family for the verification steps. The client-side
no-leader behaviour is exercised by yb-client tests
(`java/yb-client/src/test/java/org/yb/client/TestAsyncYBClient*`); the
container demo above is the public-entry-point smoke check (yb-admin
`get_universe_config` vs a quorum that cannot elect). No test-only production
APIs required.

## Confidentiality check

Artifact reviewed: contains only public GitHub issue content
(#3870/#3871/#3873/#3874), upstream source paths/lines from
yugabyte/yugabyte-db and yugabyte/charts, public Docker Hub image tags, public
PLAT-* identifiers appearing in upstream commit subjects, and synthetic
loopback addresses (127.0.0.x) from the local demo. No customer identifiers,
PII, schemas/queries, secrets, internal hostnames/URLs, or support-case
material. The issue's Aha! link is quoted from the public issue body itself.

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-issue-3873-noleader-demo.log.txt`):

=== A) leaderless quorum: master role (webserver /api/v1/masters) ===
{}
=== A) yb-admin get_universe_config vs leaderless quorum (timeout_ms=15000) ===
Timed out (yb/rpc/outbound_call.cc:839): Unable to establish connection to leader master at [127.0.0.1:7100]. Please verify the addresses and check if server is up, or if you're missing --certs_dir_name.

: Could not locate the leader master: Call timed out before sending
podman exec yb3873-noleader /home/yugabyte/bin/yb-admin -master_addresses      0.01s user 0.01s system 0% cpu 15.152 total
tail -4  0.00s user 0.00s system 0% cpu 15.151 total
=== A) master stuck without quorum (log excerpt) ===
I0830 12:25:14.869572    18 tcp_stream.cc:329] { local: 127.0.0.1:38411 remote: 127.0.0.3:7100 }:  Recv failed: Network error (yb/util/net/socket.cc:542): recvmsg error: Connection refused (system error 111) [suppressed 2401 similar messages]
W0830 12:25:15.311765    29 async_client_initializer.cc:104] Failed to initialize client: Timed out (yb/rpc/rpc.cc:224): Could not locate the leader master: GetLeaderMasterRpc(addrs: [127.0.0.1:7100, 127.0.0.2:7100, 127.0.0.3:7100], num_attempts: 52) passed its deadline 80596.855s (passed 1.526s of 1.500s): Not found (yb/master/master_rpc.cc:293): no leader found: GetLeaderMasterRpc(addrs: [127.0.0.1:7100, 127.0.0.2:7100, 127.0.0.3:7100], num_attempts: 1)
I0830 12:25:16.380888    14 tcp_stream.cc:329] { local: 127.0.0.1:32977 remote: 127.0.0.3:7100 }:  Recv failed: Network error (yb/util/net/socket.cc:542): recvmsg error: Connection refused (system error 111) [suppressed 2451 similar messages]
I0830 12:25:16.894482    14 tcp_stream.cc:329] { local: 127.0.0.1:44681 remote: 127.0.0.3:7100 }:  Recv failed: Network error (yb/util/net/socket.cc:542): recvmsg error: Connection refused (system error 111) [suppressed 2501 similar messages]
I0830 12:25:17.817099    15 tcp_stream.cc:329] { local: 127.0.0.1:40595 remote: 127.0.0.3:7100 }:  Recv failed: Network error (yb/util/net/socket.cc:542): recvmsg error: Connection refused (system error 111) [suppressed 2551 similar messages]
W0830 12:25:17.833371    29 async_client_initializer.cc:104] Failed to initialize client: Timed out (yb/rpc/rpc.cc:224): Could not locate the leader master: GetLeaderMasterRpc(addrs: [127.0.0.1:7100, 127.0.0.2:7100, 127.0.0.3:7100], num_attempts: 52) passed its deadline 80599.393s (passed 1.509s of 1.500s): Not found (yb/master/master_rpc.cc:293): no leader found: GetLeaderMasterRpc(addrs: [127.0.0.1:7100, 127.0.0.2:7100, 127.0.0.3:7100], num_attempts: 1)

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-issue-3873-control-demo.log.txt`):

=== B) control: single-master quorum, same probe ===
{"version":0,"clusterUuid":"93459171-c196-4c57-acce-e9a74bbe2643","universeUuid":"9c3bd158-9c55-4adb-a95d-94a99d63caa9"}
podman exec yb3873-leader /home/yugabyte/bin/yb-admin -master_addresses     2  0.01s user 0.01s system 10% cpu 0.163 total
head -3  0.00s user 0.00s system 0% cpu 0.162 total

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu
